aea/cli/upgrade.py

```diff
@@ -0,0 +1,315 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2018-2019 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+"""Implementation of the 'aea upgrade' subcommand."""
+from collections import defaultdict
+from pathlib import Path
+from typing import Any, Generator, Tuple, cast
+
+import click
+
+from aea.cli.add import add_item
+from aea.cli.registry.add import _get_package_meta
+from aea.cli.remove import remove_item
+from aea.cli.utils.click_utils import PublicIdParameter
+from aea.cli.utils.config import load_item_config
+from aea.cli.utils.context import Context
+from aea.cli.utils.decorators import check_aea_project, clean_after, pass_ctx
+from aea.cli.utils.package_utils import (
+    find_item_locally,
+    get_item_public_id_by_author_name,
+    get_items,
+    is_item_present,
+)
+from aea.configurations.base import AgentConfig, PackageConfiguration, PublicId
+
+
+class ItemRemoveHelper:
+    """Helper to check dependencies on removing component from agent config."""
+
+    def __init__(self, agent_config: AgentConfig) -> None:
+        """Init helper."""
+        self._agent_config = agent_config
+
+    def get_agent_dependencies(self) -> defaultdict:
+        """
+        Get all reverse dependencices in agent.
+
+        :return:  defaultdict with package id: (type, public_id) and set of component that uses this package
+        """
+        return self.get_item_dependencies(self._agent_config, None)
+
+    @staticmethod
+    def get_item_config(item_type: str, public_id: PublicId) -> PackageConfiguration:
+        """Get item config for item,_type and public_id."""
+        return load_item_config(
+            item_type,
+            package_path=Path("vendor")
+            / public_id.author
+            / f"{item_type}s"
+            / public_id.name,
+        )
+
+    @staticmethod
+    def _get_item_requirements(
+        item: Any,
+    ) -> Generator[Tuple[str, PublicId], None, None]:
+        """
+        List all the requiemenents for item provided.
+
+        :return: generator with package ids: (type, public_id)
+        """
+        item_types = ["connection", "skill", "protocol", "contract"]
```

Comment:
We already have an enum which defines them `ComponentType`

Reply:
nice point, thanks!